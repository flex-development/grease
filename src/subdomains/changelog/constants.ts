/**
 * @file Constants
 * @module grease/changelog/constants
 */

import type { ICommitType } from '#src/changelog/interfaces'
import { Type } from '@flex-development/commitlint-config'

/**
 * An array of objects representing explicitly supported commit message types,
 * and whether they should show up in the changelog.
 *
 * @see {@linkcode ICommitType}
 *
 * @const {ICommitType[]} TYPES
 */
export const TYPES: ICommitType[] = [
  { section: ':sparkles: Features', type: Type.FEAT },
  { section: ':bug: Fixes', type: Type.FIX },
  { section: ':mechanical_arm: Refactors', type: Type.REFACTOR },
  { section: ':fire: Performance Improvements', type: Type.PERF },
  { section: ':wastebasket: Reverts', type: Type.REVERT },
  { section: ':pencil: Documentation', type: Type.DOCS },
  { section: ':white_check_mark: Testing', type: Type.TEST },
  { section: ':package: Build', type: Type.BUILD },
  { section: ':robot: Continuous Integration', type: Type.CI },
  { section: ':house_with_garden: Housekeeping', type: Type.CHORE },
  { hidden: true as const, type: Type.RELEASE },
  { hidden: true as const, type: Type.STYLE },
  { hidden: true as const, type: Type.WIP }
]
