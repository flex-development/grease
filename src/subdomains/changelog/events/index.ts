/**
 * @file Entry Point - Events
 * @module grease/changelog/events
 */

export { default as ChangelogEvent } from './changelog.event'
export { default as ChangelogEventListener } from './changelog.listener'
