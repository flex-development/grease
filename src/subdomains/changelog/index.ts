/**
 * @file Entry Point - changelog
 * @module grease/changelog
 */

export { default as ChangelogModule } from './changelog.module'
export * from './constants'
export * from './events'
export type * from './interfaces'
export * from './models'
export * from './operations'
export * from './queries'
export type * from './types'
