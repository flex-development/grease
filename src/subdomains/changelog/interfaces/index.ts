/**
 * @file Entry Point - Interfaces
 * @module grease/changelog/interfaces
 */

export type { default as IChangelogEntry } from './changelog-entry.interface'
export type { default as ICommitGroup } from './commit-group.interface'
export type { default as ICommitType } from './commit-type.interface'
