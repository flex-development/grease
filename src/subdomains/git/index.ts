/**
 * @file Entry Point - git
 * @module grease/git
 */

export * from './enums'
export { default as GitModule } from './git.module'
export type * from './interfaces'
export * from './models'
export * from './operations'
export * from './options'
export * from './providers'
export * from './queries'
export type * from './types'
