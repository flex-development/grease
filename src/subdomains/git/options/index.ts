/**
 * @file Entry Point - Options
 * @module grease/git/options
 */

export { default as CommitGrammarOptions } from './commit-grammar.options'
export { default as GitOptions } from './git.options'
