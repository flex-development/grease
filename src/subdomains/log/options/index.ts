/**
 * @file Entry Point - Options
 * @module grease/log/options
 */

export { default as LoggerOptions } from './logger.options'
