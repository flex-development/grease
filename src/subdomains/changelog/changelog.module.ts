/**
 * @file ChangelogModule
 * @module grease/changelog/ChangelogModule
 */

import { LoggerService, ValidationService } from '#src/providers'
import { Module } from '@nestjs/common'
import { CqrsModule } from '@nestjs/cqrs'
import { ChangelogEventListener } from './events'
import { ChangelogOperationHandler } from './operations'
import { ChangelogQueryHandler } from './queries'

/**
 * Changelog module.
 *
 * @class
 */
@Module({
  imports: [CqrsModule],
  providers: [
    ChangelogEventListener,
    ChangelogOperationHandler,
    ChangelogQueryHandler,
    LoggerService,
    ValidationService
  ]
})
class ChangelogModule {}

export default ChangelogModule
