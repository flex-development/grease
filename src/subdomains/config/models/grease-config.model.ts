/**
 * @file Models - GreaseConfig
 * @module grease/config/models/GreaseConfig
 */

import { ChangelogOperation } from '#src/changelog'
import type { IGreaseConfig } from '#src/config/interfaces'
import { TagOperation, type Commit } from '#src/git'
import { GlobalOptions } from '#src/options'
import { define, get, keys, set } from '@flex-development/tutils'
import { IsInstance, ValidateNested } from 'class-validator'

/**
 * Grease configuration options.
 *
 * @see {@linkcode GlobalOptions}
 * @see {@linkcode IGreaseConfig}
 *
 * @template T â€” Parsed commit type
 *
 * @class
 * @extends {GlobalOptions}
 * @implements {IGreaseConfig<T>}
 */
class GreaseConfig<T extends Commit = Commit> extends GlobalOptions
  implements IGreaseConfig<T> {
  /**
   * Changelog options.
   *
   * @see {@linkcode ChangelogOperation}
   *
   * @default
   *  new ChangelogOperation(opts.changelog)
   *
   * @public
   * @instance
   * @member {ChangelogOperation<T>} changelog
   */
  @IsInstance(ChangelogOperation)
  @ValidateNested()
  public changelog: ChangelogOperation<T>

  /**
   * Tag options.
   *
   * @see {@linkcode TagOperation}
   *
   * @default
   *  new TagOperation(opts.tag)
   *
   * @public
   * @instance
   * @member {TagOperation} tag
   */
  @IsInstance(TagOperation)
  @ValidateNested()
  public tag: TagOperation

  /**
   * Create a new config object.
   *
   * @see {@linkcode IGreaseConfig}
   *
   * @param {IGreaseConfig<T>?} [opts={}] - User config
   */
  constructor(opts: IGreaseConfig<T> = {}) {
    super(opts)

    define(opts, 'changelog', { value: get(opts, 'changelog', {}) })
    define(opts, 'tag', { value: get(opts, 'tag', {}) })

    Reflect.deleteProperty(opts.changelog!, 'from')
    Reflect.deleteProperty(opts.tag!, 'tag')

    for (const key of keys(opts)) {
      switch (key) {
        case 'changelog':
        case 'tag':
          set(opts, key, { ...get(opts, key), ...this })
          break
        default:
          break
      }
    }

    this.changelog = new ChangelogOperation<T>(opts.changelog)
    this.tag = new TagOperation({ ...opts.tag, tag: '' })
  }
}

export default GreaseConfig
