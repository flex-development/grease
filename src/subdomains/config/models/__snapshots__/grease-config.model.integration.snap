// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`integration:config/models/GreaseConfig > validation > should fail validation if schema is invalid 1`] = `
[
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isBoolean": "colors must be a boolean value",
        },
        "property": "colors",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isFile": "config must be a file path or url",
        },
        "property": "config",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isBoolean": "debug must be a boolean value",
        },
        "property": "debug",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isBoolean": "quiet must be a boolean value",
        },
        "property": "quiet",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isString": "tagprefix must be a string",
        },
        "property": "tagprefix",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "min": "releases must not be less than -1",
        },
        "property": "releases",
      },
    ],
    "property": "changelog",
  },
  ValidationError {
    "children": [
      ValidationError {
        "children": [],
        "constraints": {
          "isBoolean": "colors must be a boolean value",
        },
        "property": "colors",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isFile": "config must be a file path or url",
        },
        "property": "config",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isBoolean": "debug must be a boolean value",
        },
        "property": "debug",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isBoolean": "quiet must be a boolean value",
        },
        "property": "quiet",
      },
      ValidationError {
        "children": [],
        "constraints": {
          "isString": "tagprefix must be a string",
        },
        "property": "tagprefix",
      },
    ],
    "property": "tag",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "colors must be a boolean value",
    },
    "property": "colors",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isFile": "config must be a file path or url",
    },
    "property": "config",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "debug must be a boolean value",
    },
    "property": "debug",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isBoolean": "quiet must be a boolean value",
    },
    "property": "quiet",
  },
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "tagprefix must be a string",
    },
    "property": "tagprefix",
  },
]
`;
